File: mobile/app/src/main/java/ru/morozovit/recipeviewer/mobile/RecipeDatabase.java
package ru.morozovit.recipeviewer.mobile;

import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import androidx.annotation.NonNull;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

public final class RecipeDatabase {
    private static RecipeDatabase INSTANCE = null;
    private final SharedPreferences sharedPreferences;

    public static class StorageException extends Exception {
        public StorageException(String message) {
            super(message);
        }
    }

    private RecipeDatabase(SharedPreferences sharedPreferences) {
        this.sharedPreferences = sharedPreferences;
    }

    /** @noinspection deprecation*/
    public static RecipeDatabase getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new RecipeDatabase(PreferenceManager.getDefaultSharedPreferences(MainActivity.getInstance().getApplicationContext()));
        }
        return INSTANCE;
    }

    public void saveRecipe(@NonNull JSONObject recipe) throws JSONException {
        String key = recipe.getString("name");
        sharedPreferences.edit().putString(key, recipe.toString()).apply();
    }

    @NonNull
    public JSONObject getRecipe(@NonNull String name) throws StorageException, JSONException {
        JSONObject obj = new JSONObject(sharedPreferences.getString(name, "{}"));
        if (!(obj.has("name") && obj.has("body"))) {
            throw new StorageException(String.format("Recipe with name '%s' does not exist", name));
        }
        return obj;
    }

    public void deleteRecipe(@NonNull String name) {
        sharedPreferences.edit().remove(name).apply();
    }

    @NonNull
    public ArrayList<JSONObject> getAllRecipes() {
        ArrayList<JSONObject> recipes = new ArrayList<>();
        for (String key : sharedPreferences.getAll().keySet()) {
            try {
                recipes.add(getRecipe(key));
            } catch (Exception e) {}
        }
        return recipes;
    }
}

File: mobile/app/src/main/java/ru/morozovit/recipeviewer/mobile/RecipeDatabase.kt
@file:Suppress("DEPRECATION")

package ru.morozovit.recipeviewer.mobile

import android.content.SharedPreferences
import android.preference.PreferenceManager
import org.json.JSONException
import org.json.JSONObject

object RecipeDatabase {
    private lateinit var sharedPreferences: SharedPreferences

    class StorageException(message: String) : Exception(message)

    private fun init() {
        if (!::sharedPreferences.isInitialized) {
            sharedPreferences = PreferenceManager.getDefaultSharedPreferences(MainActivity.getInstance().applicationContext)
        }
    }

    fun saveRecipe(recipe: JSONObject) {
        init()
        val key = recipe.getString("name")
        sharedPreferences.edit().putString(key, "$recipe").apply()
    }

    @Throws(JSONException::class, StorageException::class)
    fun getRecipe(name: String): JSONObject {
        init()
        val obj = JSONObject(sharedPreferences.getString(name, "{}")!!)
        if (!(obj.has("name") && obj.has("body"))) {
            throw StorageException(String.format("Recipe with name '%s' does not exist", name))
        }
        return obj
    }

    fun deleteRecipe(name: String) {
        init()
        sharedPreferences.edit().remove(name).apply()
    }

    fun getAllRecipes(): ArrayList<JSONObject> {
        init()
        val recipes = ArrayList<JSONObject>()
        for (key in sharedPreferences.all.keys) {
            try {
                recipes.add(getRecipe(key))
            } catch (_: Exception) {}
        }
        return recipes
    }
}